crates/bevy_reflect/src/from_reflect.rs

```diff
@@ -12,7 +13,29 @@ use crate::{FromType, Reflect};
 /// reflected value.
 pub trait FromReflect: Reflect + Sized {
     /// Constructs a concrete instance of `Self` from a reflected value.
-    fn from_reflect(reflect: &dyn Reflect) -> Option<Self>;
+    fn from_reflect(reflect: &dyn Reflect) -> Result<Self, FromReflectError>;
+}
+
+/// An enum representing all the kinds of types that a value implementing [`Reflect`] could be.
+/// This enum is primarily used in [`FromReflectError`].
+#[derive(Debug)]
+pub enum ReflectType {
+    Struct,
+    TupleStruct,
+    Tuple,
+    List,
+    Array,
+    Map,
+    Enum,
+    Value,
+}
+
+/// An Error for failed conversion of reflected type to original type in [`FromReflect::from_reflect`].
+#[derive(Error, Debug)]
+#[error("The reflected type of `{}` cannot be converted to {:?}", .from_type_name, .to_type)]
+pub struct FromReflectError<'a> {
+    pub from_type_name: &'a str,
```

Comment:
One concern here: it might be important to know the actual type of the given value as well. For example, we may want to know that we're actually passing a `DynamicStruct` with the type name `foo::Foo` rather than an actual instance of `foo::Foo`.

Would it make sense to store the "target" type name as well as the `std::any::type_name` of the passed value?